### DISCUSSION TITLE
General waveuguide port support in openEMS

### DISCUSSION THREAD

**Discussion by gadiLhv** [thliebig/openEMS-Project#188] [Show and tell]
So here is a project that has been going on for quite a while.

We wanted to enable support for any sort of waveguide port in openEMS. The strategy was pretty straight forward:

1. Solve the eigenmodes for the waveguide structure using another open-source FEM library.
2. Extract the E and H values on the FDTD mesh grid, along with the propagation coefficients and mode impedance. I didn't figure out line impedance yet (characteristic impedance).
3. Use the existing E\H excitations as ports and manually load said values
4. Same with the E\H field probes, manually load the values.

This strategy allowed minimal changes to the existing openEMS functionality (and source core, for that matter).

However, in order to be able to place the ports anywhere **within** the solved domain, it was necessary to implement some sort of absorber within the domain, rather than near the edges of the air box.

We have a working implementation for 1st order Mur and 1st order Mur with super-absorption, which I discussed previously [here](https://github.com/thliebig/openEMS-Project/discussions/107#discussioncomment-6982148). I know that surface impedance BC (**SIBC**) are rather popular also, but I'll discuss that a bit later. 

### Microstrip Line
Since it won't really be interesting to share the rectangular waveguide ports, here is the microstrip line mode calculated using the FEM library, 

<img src="https://github.com/thliebig/openEMS-Project/assets/3908295/5bd143bc-5783-4b43-b63b-5fc6d9b7a2ce" width = "550">

Then sampled at the FDTD grid:

<img src="https://github.com/thliebig/openEMS-Project/assets/3908295/05a8add3-4468-41a8-8539-fe3cb67d2a6d" width="550">

Placed two ports, one exciting and one receiving, along with absorbers behind each one. The phase velocity of the Mur B.C. was set according to the calculated propagation coefficient. The absorber is set at least one mesh cell **behind** the excitation, in the current implementation. It can be seen that the structure is within a larger airbox. In this simulation Mur BCs were used around the airbox.

<img src = "https://github.com/thliebig/openEMS-Project/assets/3908295/a40536ce-d082-446e-84b1-b33068dca333" width = "450">

https://github.com/thliebig/openEMS-Project/assets/3908295/b173c1d8-7feb-44a2-882e-f1dafda3c165

https://github.com/thliebig/openEMS-Project/assets/3908295/e8c40830-f2a0-46ba-aa0a-df9e136edf7b

The S-Parameters are currently calculated using the wave impedance. Line impedance is a bit more tricky using the current FEM library.

<img src = "https://github.com/thliebig/openEMS-Project/assets/3908295/c2dcb546-da8a-43d2-bbc5-c6ce9685b7ff" width ="450">

### Future work
Although adding "custom" modes to openEMS required relatively little changes, the "local" absorbers required quite a bit. Additional modules and definitions for CSXCAD, another engine extension and a bit of hacking to support multithreading, which is why I started [this issue](https://github.com/thliebig/openEMS/issues/122) back in the day.

Intuitively, I have a feeling that a more healthy approach would be to somehow absorb specifically the excited mode, although I didn't figure that out yet. I found a few articles in the field, referenced below, but couldn't figure it out yet.

@thliebig, do you have any notion if it's possible to excite and absorb reflected modes on the same reference plane? I couldn't exactly figure out the articles I found (referenced below), but it seems the fields are sampled close to the excitation. 

I am asking this, as in commercial software, it is commonly done with metallic shielding behing the port. This could, of course, be a visually pleasing approach for a friendly user interface (and maintain consistency with FEM waveguide ports), but I have only guesses at the moment. 

For example:
<img src="https://github.com/thliebig/openEMS-Project/assets/3908295/b77246ac-e3c4-49c7-9dbd-7d53369ed622" width = "400">

I feel like it will be somewhat more efficient if I could add this functionality to the excitation extension, rather than define a separate module of a "generic" absorber. Maybe it's just enough to add **SIBC** functionality, but it would be excellent to hear an expert opinion before delving back in to trial and error.

Would love to hear your input on this.

References:
[1] ALIMENTI, Federico, et al. A revised formulation of modal absorbing and matched modal source boundary conditions for the efficient FDTD analysis of waveguide structures. IEEE Transactions on Microwave Theory and Techniques, 2000, 48.1: 50-59.‏

[2] ALIMENTI, F., et al. Modal absorption in the FDTD method: A critical review. International Journal of Numerical Modelling: Electronic Networks, Devices and Fields, 1997, 10.4: 245-264.‏

**Comment by toammann** [comment]
Hello gadiLhv,
thank you very much for your effort in implementing this! I think having a port mode solver of a line geometry can be very useful (also for debugging purposes). With this technique, openEMS should support arbitrary transmission line geometries.

A few questions on the implementation:
 - Just out of curiosity: What open source library do you use for the port mode solver?
 - Does this support dispersive (Lorentz/Drude) materials?
 - Due to the known propagation constant: Is a de-embedding feature (reference plane shift) possible?

Currently, I do not know how accurate the openEMS microstrip port is if the geometry differs from the standard mode. Let's say a shielding GND on the top layer is near the RF line. In this case a portion of the return current will flow on the top layer forming a GCWG. The distances to GND on the top layer may even be asymetrical. The port mode solver should treat this accurately. Does the current implementation of the mircostrip port in openEMS, too? I think the recorded reflection coefficient may be correct (independent on line impedance), but I don´t know if the line impedance is. 

Again, tanks for your efforts!

Regards,
Tobias

**Reply by gadiLhv** [reply]
Hey

>Just out of curiosity: What open source library do you use for the port mode solver?

I don't want to share that right now. It's a work in progress, to say the least. The library I currently had success with has prolematic formulation and I had to brute force some stuff. I don't think it's very deductive... I am pushing that the end result will be in [FEniCS](https://fenicsproject.org/). 

I am willing to share the reference I used to get to the weak form formulation for waveguide modes, if you are interested.

>Does this support dispersive (Lorentz/Drude) materials?

Pretty sure it does. Didn't get there yet. Getting this far was terrible, on it's own. I really thought I knew a thing or two about FEM. This one kicked my ass, real good. Still does, BTW.

>Due to the known propagation constant: Is a de-embedding feature (reference plane shift) possible?

Absolutely, but you will need to calculate the propagation coefficient in several frequecnies, to interpolate it properly. A single frequency calculation will not suffice in a wide band calculation.

>Currently, I do not know how accurate the openEMS microstrip port is if the geometry differs from the standard mode. Let's say a shielding GND on the top layer is near the RF line. In this case a portion of the return current will flow on the top layer forming a GCWG. The distances to GND on the top layer may even be asymetrical.

Yup. This was one of the reasons to start this process.

>The port mode solver should treat this accurately

It will calculate it a better, especially in high frequency applications. There is an issue with an E and H pair calculated in a single frequency, for a time-domain simulation. If you are not familiar with this, it is very important to know. Up until a certain bandwidth it should be Ok.

>Does the current implementation of the mircostrip port in openEMS, too?  I think the recorded reflection coefficient may be correct (independent on line impedance), but I don´t know if the line impedance is.

I don't understand the question, sorry. I have to say I never delved into the MSL port in openEMS.

**Comment by GitOA7** [comment]
Hello, There is a very simple method to get the modes in FDTD. You can run a small FDTD simulation only in the box used for the port by using MUR (or may be PMC is better in this case ) conditions and by initializing by a random Electric or magnetic field. At the end of the simulation, using an FFT on a saved field will give the modes of the structure by detecting the pics. The first pic corresponds to the first mode. This field can be used for the excitation. There are many papers about mode calculation using FDTD. This method is usually used to calculate dispersion diagram for periodic structures. Actually, it can be used to calculate the modes of any structure. Here is a paper describing this method for periodic structures : "FDTD Modeling of Periodic Structures: A Review" ; link : https://arxiv.org/pdf/2007.05091 . Hope this can help. ++

**Reply by gadiLhv** [reply]
If I understand correctly, you suggested a way to obatin the mode itself, not excite the mode in the simulation, correct?

**Comment by GitOA7** [comment]
Yes, the method will give the mode its self (electric field and magnetic field values at each point of the mesh), not the excitation. And I think, the best configuration is to use PMC conditions. The filed will configure itself depending on the structure and the materials inside the port box. For the excitation, I don't know how we can tell openems to use this field. But I suppose it can be a waveguide like port and we have to find a way to impose the field calculated by the small FDTD simulation that we have run only inside the port box ( for example, by considering the predefined mesh grid for the big simulation  and by running the simulation only inside the port box using a ~10% of the number of iterations defined by the user).

**Reply by gadiLhv** [reply]
Hey @GitOA7

For the modes solver, I'd much rather use an FEM solver, as I did.
Maybe an IE solver is an option as well, just that FEM solvers are readily available, open source. 

Cheers
